### ISSUE TITLE
Doesn't work with Citrix 21.6.0

### ISSUE THREAD

**Issue by chris-day** [andyjmorgan/CaffeineForCitrixWorkspace#4]
The registry keys are different in this version. By adding the keys the application doesn't work as expected.

**Comment by P-rox-y** [comment]
Seeing the same behavior. I'm on Citrix 21.8.x and it looks like the only keys are

\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Citrix\ICA Client

\CEIP
\Engine
\GenericUSB
\Prelaunch

**Comment by ccz245** [comment]
Same problem here, would be great if we can get a fix for this

**Comment by eXpl0it3r** [comment]
See a PowerShell script workaround in #2

**Comment by badbob001** [comment]
What is the last release of Citrix Workspace does this still work on? Or do we have to go way back to Citrix Receiver 4.9?

**Comment by elboertjie** [comment]
> What is the last release of Citrix Workspace does this still work on? Or do we have to go way back to Citrix Receiver 4.9? Even with 4.9, the CCM key needs to be added manually.
> 
> Update: tried 2102 and it did not work (right-clicking icon would not produce menu, indicating it has frozen). Tried 2010 and it works.

The original Caffeine works with the latest Citrix Workspace 2202.  There was initially an error message for an integer value too small or too large.  Simply uninstall everything, re-install Citrix and then re-install Caffeine (the original one).

**Comment by badbob001** [comment]
> What trick did you use with the PowerShell?

I used the powershell script from [here](https://github.com/andyjmorgan/CaffeineForCitrixWorkspace/issues/2), so all thanks to the people in that thread.

I commented out the while loop part so it just runs once. Then I schedule it to run in task scheduler. Below, replace <script path> with actual path of the script files. The invisible.vbs hides the script window from showing on the desktop when run. Probably better ways to implement this and with fewer extra files, but I can add other commands to the cmd file if I want them to repeat on the same schedule.

I also modified the powershell script to output to a log file so I can verify that it can see my citrix sessions, but would be nicer if I can expose that information to a tasktray icon. Perhaps moving the loop from task scheduler to autohotkey?

Task:
> Program: %systemroot%\system32\wscript.exe
> Arguments: "invisible.vbs" "repeat_nonadmin.cmd" //nologo
> Start in: <script path>
> Run only when user is logged on
> DISABLED: Run with highest privileges
> Trigger: Run at 12am and repeat every hour

invisible.vbs:
`CreateObject("Wscript.Shell").Run """" & WScript.Arguments(0) & """", 0, False`

repeat_nonadmin.cmd:
`c:\windows\syswow64\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -File "<script path>\keep_ica_alive.ps1"`

keep_ica_alive.ps1:
```
$logfile = '<script path>' + "\keep_ica_alive_" + $(get-date -f yyyyMM) + ".log"
[System.Reflection.Assembly]::LoadFile("C:\Program Files (x86)\Citrix\ICA Client\WfIcaLib.dll") | Out-Null;
$ICO = New-Object WFICALib.ICAClientClass;
$ICO.OutputMode = [WFICALib.OutputMode]::OutputModeNormal;
$EnumHandle = $ICO.EnumerateCCMSessions();
$NumSessions = $ICO.GetEnumNameCount($EnumHandle);

for ($index = 0; $index -lt $NumSessions; $index++)
{
    $sessionid = $ICO.GetEnumNameByIndex($EnumHandle, $index)
    $ICO.StartMonitoringCCMSession($sessionid, $true);
    $ICO.Session.Keyboard.SendKeyDown(126);
    $ICO.StopMonitoringCCMSession($sessionid);
}

$ICO.CloseEnumHandle($EnumHandle) | Out-Null;

Write-output "$(get-date -f 'yyyy/MM/dd HH:mm') Active sessions: $NumSessions"
Write-output "$(get-date -f 'yyyy/MM/dd HH:mm') Active sessions: $NumSessions" | out-file -Append -FilePath $logfile
if(Resolve-AzError) {
    Resolve-AzError | out-file -Append -FilePath $logfile
}

```

**Comment by PTCruiserGT** [comment]
I know 

> > The original Caffeine works with the latest Citrix Workspace 2202. There was initially an error message for an integer value too small or too large. Simply uninstall everything, re-install Citrix and then re-install Caffeine (the original one).
> 
> Couldn't test this since I can't get the original caffeine to work on my system. Just running it would get an access denied error. And it was even harder to uninstall it because I get an error when it tries to access HKCU\Software\Microsoft\SystemCertificates\Root\ProtectedRoots. Hmm, perhaps I got a malware version (couldn't download from the Box.com link in the original site). Or older buggy version? Version says 1.0.9.
> 
> I did manage to get a powershell method to work on the latest release so I think I'm still covered.

I wanted to mention that the "original" Caffeine also does not work for me, but version 1.2 that Andrew released in 2015 does work.

**Comment by eXpl0it3r** [comment]
Yeah, it seems whatever caused the hanging in the DLL class instantiation was "fixed", so it actually started working again

**Comment by TTTrouble** [comment]
Had to hunt around for the original 1.2 Caffeine download link and thought I'd share for anyone else that's looking. The link for the 1.2 version cited above is buried in the comments of the original post, link from Andrew's comment - https://app.box.com/s/k45pz5vyhnp5ejbr474pb0up117giduh
